rosetta/retriever/retriever_test.go

```diff
@@ -0,0 +1,1066 @@
+				return nil
+			},
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, dps.FlowSymbol, symbol)
+				return "deposit", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, dps.FlowSymbol, symbol)
+				return "", mocks.DummyError
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+		}
+
+		_, err := r.Transaction(testBlockID, testTransactionID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles index retrieval failures", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				assert.Equal(t, testBlockID, block)
+				return block, nil
+			},
+			TransactionFunc: func(transaction identifier.Transaction) error {
+				assert.Equal(t, testTransactionID, transaction)
+				return nil
+			},
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, dps.FlowSymbol, symbol)
+				return "deposit", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, dps.FlowSymbol, symbol)
+				return "withdrawal", nil
+			},
+		}
+		index := &mocks.Reader{
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				assert.Equal(t, testHeight, height)
+				if assert.Len(t, types, 2) {
+					assert.Equal(t, flow.EventType("deposit"), types[0])
+					assert.Equal(t, flow.EventType("withdrawal"), types[1])
+				}
+
+				return nil, mocks.DummyError
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			index:     index,
+		}
+
+		_, err := r.Transaction(testBlockID, testTransactionID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles incorrectly-formatted indexed events", func(t *testing.T) {
+		invalidEvents := []flow.Event{
+			{
+				Payload: []byte(`invalid_format`),
+			},
+		}
+
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				assert.Equal(t, testBlockID, block)
+				return block, nil
+			},
+			TransactionFunc: func(transaction identifier.Transaction) error {
+				assert.Equal(t, testTransactionID, transaction)
+				return nil
+			},
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, dps.FlowSymbol, symbol)
+				return "deposit", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, dps.FlowSymbol, symbol)
+				return "withdrawal", nil
+			},
+		}
+		index := &mocks.Reader{
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				assert.Equal(t, testHeight, height)
+				if assert.Len(t, types, 2) {
+					assert.Equal(t, flow.EventType("deposit"), types[0])
+					assert.Equal(t, flow.EventType("withdrawal"), types[1])
+				}
+
+				return invalidEvents, nil
```

Comment:
Rather than using `invalidEvents`, as suggested earlier, we should separate out the conversion functionality and mock it, so we don't have to make any assumptions about event formats. We just make the converter error when we want, and successfully return the correct operations when we want.